pulsar-broker/src/test/java/org/apache/pulsar/client/api/DeduplicationEndToEndTest.java

```diff
@@ -0,0 +1,672 @@
+        PulsarCommandSender spyCommandSender = Mockito.spy(commandSender);
+        commandSenderField.set(serverCnx, spyCommandSender);
+
+        // disable the send receipt
+        Mockito.doNothing().when(spyCommandSender).sendSendReceiptResponse(Mockito.anyLong(), Mockito.anyLong(),
+                Mockito.anyLong(), Mockito.anyLong(), Mockito.anyLong());
+        return commandSender;
+    }
+
+    /**
+     * Set the original commandSender back to the ServerCnx, so that the producer can receive the ack.
+     * @param topic
+     * @param producerName
+     * @param sender
+     * @throws Exception
+     */
+    private void enableSendReceipt(String topic, String producerName, PulsarCommandSender sender) throws Exception {
+        PersistentTopic persistentTopic = (PersistentTopic) pulsar.getBrokerService()
+                .getTopic(topic + "-partition-" + 0, false).get().get();
+        assertNotNull(persistentTopic);
+        ServerCnx serverCnx = (ServerCnx) persistentTopic.getProducers().get(producerName).getCnx();
+
+        // set original commandSender back
+        Field commandSenderField = ServerCnx.class.getDeclaredField("commandSender");
+        commandSenderField.setAccessible(true);
+        commandSenderField.set(serverCnx, sender);
+    }
+
+    /**
+     * simulate the case where the producer sends the message but doesn't receive the ack
+     * due to network issue, broker issue, etc. User receives the exception and sends the same message again.
+     * The message is duplicated in the topic.
+     * @throws Exception
+     */
+    @Test
+    public void testProducerDuplicationWithReceiptLost() throws Exception {
+        final String topic = "persistent://my-property/my-ns/deduplication-test";
+        int partitionCount = 1;
+        admin.topics().createPartitionedTopic(topic, partitionCount);
+        admin.namespaces().setDeduplicationStatus("my-property/my-ns", false);
+
+        // Create producer with deduplication disabled
+        String producerName = "my-producer-name";
+        Producer<byte[]> producer = pulsarClient.newProducer().topic(topic)
+                .producerName(producerName).sendTimeout(1, TimeUnit.SECONDS).create();
+        assertEquals(producer.getLastSequenceId(), -1L);
+        Consumer<byte[]> consumer = pulsarClient.newConsumer().topic(topic).subscriptionName("my-sub").subscribe();
+
+        // disable the send receipt to simulate the case where the producer sends the message but doesn't receive the ack
+        PulsarCommandSender sender = disableSendReceipt(topic, producerName);
+
+        // send a message
+        byte[] data = "test".getBytes();
+        producer.sendAsync(data).thenRun(() -> {
+            // should not enter here
+            Assert.fail();
+        }).exceptionally(e -> {
+            // do not receive the ack, should enter here
+            return null;
+        }).get();
```

Comment:
```suggestion
        Assert.assertThrows(PulsarClientException.TimeoutException.class, () -> producer.send(data));
```

You'd better assert the specific exception that it could throw. BTW, synchronous `send` is simple here.

Reply:
Good idea, thanks!